lib/systems/parse.nix

```diff
@@ -496,7 +496,15 @@ rec {
                           gnuNetBSDDefaultExecFormat cpu != kernel.execFormat)
         kernel.execFormat.name;
     optAbi = lib.optionalString (abi != abis.unknown) "-${abi.name}";
-  in "${cpu.name}-${vendor.name}-${kernelName kernel}${optExecFormat}${optAbi}";
+  in
+    # gnu-config considers "mingw32" and "cygwin" to be kernels.
+    # This is obviously bogus, which is why nixpkgs has historically
+    # parsed them differently.  However for regression testing
+    # reasons (see lib/tests/triples.nix) we need to replicate this
+    # quirk when unparsing in order to round-trip correctly.
+    if      abi == abis.cygnus        then "${cpu.name}-${vendor.name}-cygwin"
+    else if kernel == kernels.windows then "${cpu.name}-${vendor.name}-mingw32"
+    else "${cpu.name}-${vendor.name}-${kernelName kernel}${optExecFormat}${optAbi}";
```

Comment:
I think this *should* be possible to eliminate. You changed the normal form so the config we used in in `examples.nix` is in normal form. I would like to instead start migrating us to use `...-windows-gnu` for the `config` in `examples`, and keep `...-windows-gnu` as the normal form.

Reply:
Does your testing current use `examples.nix`? I think we can just skip the `...-{mingw32,cygwin}` test cases but to test `...-windows-gnu` and `...-windows-msvc` to be able to land something before the migration of `examples.nix` is complete.

Reply:
> I think this _should_ be possible to eliminate.

It's not.  Your patch to GNU config made `msvc` and `gnu` valid windows ABIs, but did not make `cygnus` a valid ABI.  So we either have to keep doing the parse-time translation or else make a breaking change to anybody out-of-tree who's referencing `lib.systems.abis.cygnus`.  If you want to submit a PR to do the deprecation+migration for `abis.cygnus` after this one I'll review it, but it will be tricky and this PR is already quite large.

> You changed the normal form so the config we used in in `examples.nix` 

> Does your testing current use `examples.nix`?

No.  This PR has nothing to do with `examples.nix`.

The test suite tests the parser (`lib.systems.parser.mkSystemFromSkeleton`).  The parser is used by lots of things.  Most things that use `examples.nix` (like `pkgsCross`) also use the parser; that's the only connection between the two.

Reply:
https://github.com/llvm/llvm-project/commit/edbdd2e5df8b59dac8ae5f45059407f8a79850d6 Ah I see, yes I only did 2 of these.

Reply:
So concretely, what I am proposing is that we remove this change, and instead filter out `windows-cygnus` from the test suite until the patch is merged. That would also keep things passing, right?

Reply:
https://lists.gnu.org/archive/html/config-patches/2023-08/msg00011.html here is the patch.